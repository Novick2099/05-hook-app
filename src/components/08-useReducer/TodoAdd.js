import React from 'react';
import {Button, Input, Typography} from "@mui/material";
import {useForm} from "../../hooks/useForm";

const TodoAdd = ({handleAddTodo}) => {

  const [{description}, handleInputChange, reset] = useForm({
    description: ''
  })

  const handleSubmit = (event)=>{
    event.preventDefault();
    if(description.trim().length <= 1){
      return;
    }
    const newTodo = {
      id: new Date().getTime(),
      desc: description,
      done: false,
    };
    const action = {
      type: 'add',
      payload: newTodo,
    };
    handleAddTodo( action );
    reset();
  }

  return (
    <>
      <Typography variant="h3">Add Todo</Typography>
      <form onSubmit={handleSubmit}>
        <Input
          type="text"
          name="description"
          placeholder="Aprender ..."
          autoComplete="off"
          value={description}
          onChange={handleInputChange}
        />
        <Button type="submit" variant="contained" color="primary">
          Add
        </Button>
      </form>

    </>
  );
};

export default TodoAdd;
