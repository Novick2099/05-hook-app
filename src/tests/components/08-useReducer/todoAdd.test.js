

import TodoAdd from "../../../components/08-useReducer/TodoAdd";
import {mount} from "enzyme";

describe('Test in todoAdd',() => {

  const handleAddTodo = jest.fn();
  const wrapper = mount(<TodoAdd
    handleAddTodo={handleAddTodo}
  />)

  it('Should match with snapshot',()=>{

    expect(wrapper).toMatchSnapshot();
});

  it('DonÂ´t should call handleAddTodo',()=>{
    // todo llamar a una funcion que es una propiedad de un objeto en un test
    const formSubmit = wrapper.find('form').prop('onSubmit');
    // todo un evento es una funcion??
    formSubmit({ //todo estoy pasando la funcion como prop del form ??
      preventDefault() {
      }
    });
    expect(handleAddTodo).toHaveBeenCalledTimes(0);

  });

  it('Should call handleAddTodo',()=>{
    const value = 'Aprender ApolloClient';

    //todo how to simulate a change in an input
    wrapper.find('input').simulate('change',{
      target: {
        value: value,
        name: 'description'
      }
    });
    const formSubmit = wrapper.find('form').prop('onSubmit');
    formSubmit({ preventDefault(){}});

    const actionObj = {
      type: "add",
        payload: {
          id: expect.any(Number),
          desc:value,
          done:false,
        }
    }

    expect( handleAddTodo ).toHaveBeenCalledTimes(1);
    expect( handleAddTodo ).toHaveBeenCalledWith(expect.any(Object));
    expect( handleAddTodo ).toHaveBeenCalledWith(actionObj);

    //No se esta reseteando
    //expect( wrapper.find('input').prop('value')).toBe('');

  });





});