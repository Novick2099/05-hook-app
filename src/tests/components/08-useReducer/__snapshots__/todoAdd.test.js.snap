// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Test in todoAdd Should match with snapshot 1`] = `
Array [
  Array [
    "",
    <h3
      className="MuiTypography-root MuiTypography-h3 css-gepadz-MuiTypography-root"
    >
      Add Todo
    </h3>,
  ],
  <form
    onSubmit={[Function]}
  >
    Array [
      "",
      Array [
        "",
        <div
          className="MuiInput-root MuiInput-underline MuiInputBase-root MuiInputBase-colorPrimary css-a88p61-MuiInputBase-root-MuiInput-root"
          onClick={[Function]}
        >
          Array [
            "",
            <input
              autoComplete="off"
              className="MuiInput-input MuiInputBase-input css-1x51dt5-MuiInputBase-input-MuiInput-input"
              name="description"
              onAnimationStart={[Function]}
              onBlur={[Function]}
              onChange={[Function]}
              onFocus={[Function]}
              placeholder="Aprender ..."
              type="text"
              value=""
            />,
          ]
        </div>,
      ],
    ]
    Array [
      "",
      Array [
        "",
        <button
          className="MuiButton-root MuiButton-contained MuiButton-containedPrimary MuiButton-sizeMedium MuiButton-containedSizeMedium MuiButtonBase-root  css-sghohy-MuiButtonBase-root-MuiButton-root"
          disabled={false}
          onBlur={[Function]}
          onContextMenu={[Function]}
          onDragLeave={[Function]}
          onFocus={[Function]}
          onKeyDown={[Function]}
          onKeyUp={[Function]}
          onMouseDown={[Function]}
          onMouseLeave={[Function]}
          onMouseUp={[Function]}
          onTouchEnd={[Function]}
          onTouchMove={[Function]}
          onTouchStart={[Function]}
          tabIndex={0}
          type="submit"
        >
          Add
          Array [
            "",
            <span
              className="MuiTouchRipple-root css-8je8zh-MuiTouchRipple-root"
            />,
          ]
        </button>,
      ],
    ]
  </form>,
]
`;
